lib/Core/BuildEngine.cpp

```diff
@@ -931,31 +942,97 @@ class BuildEngineImpl {
         didWork = true;
       }
 
-      // If we didn't do any work, we are done.
-      if (!didWork)
-        break;
-    }
+      if (!didWork) {
+        // If there was no work to do, but we still have running tasks, then
+        // we have found a cycle. Try to resolve it and continue.
+        if (!taskInfos.empty()) {
+          if (resolveCycle(buildKey)) {
+            continue;
+          } else {
+            cancelRemainingTasks();
+            return false;
+          }
+        }
 
-    // If there was no work to do, but we still have running tasks, then
-    // we have found a cycle and are deadlocked.
-    if (!taskInfos.empty()) {
-      reportCycle(buildKey);
-      return false;
+        // We didn't do any work, and we have nothing more we can/need to do.
+        break;
+      }
     }
 
     return true;
   }
 
-  /// Report the cycle which has called the engine to be unable to make forward
+  /// Attempt to resolve a cycle which has called the engine to be unable to make forward
   /// progress.
   ///
   /// \param buildKey The key which was requested to build (the reported cycle
   /// with start with this node).
-  void reportCycle(const KeyType& buildKey) {
+  /// \returns True if the engine should try to proceed, false if the build the could not
+  /// be broken.
+  bool resolveCycle(const KeyType& buildKey) {
     // Take all available locks, to ensure we dump a consistent state.
     std::lock_guard<std::mutex> guard1(taskInfosMutex);
     std::lock_guard<std::mutex> guard2(finishedTaskInfosMutex);
 
+    std::vector<Rule*> cycleList = findCycle(buildKey);
+    assert(!cycleList.empty());
+
+    if (delegate.shouldResolveCycle(cycleList)) {
+      TracingInterval i(EngineQueueItemKind::BreakingCycle);
+
+      for (auto ruleIt = cycleList.rbegin(); ruleIt != cycleList.rend(); ruleIt++) {
+        auto& ruleInfo = getRuleInfoForKey((*ruleIt)->key);
+
+        // If this rule is scanning, try to force a rebuild to break the cycle
+        if (ruleInfo.isScanning()) {
+          if (trace) {
+            trace->cycleForceRuleNeedsToRun(&ruleInfo.rule);
+          }
+          for (auto it = ruleInfosToScan.begin(); it != ruleInfosToScan.end(); it++) {
+            if (it->ruleInfo->keyID == ruleInfo.keyID) {
+              ruleInfosToScan.erase(it);
```

Comment:
This should break, I think (I think it can actually crash if it doesn't)?

Actually, maybe we should probably find the node, then assert we found it, then erase (same comment below)

Reply:
I think you can also just use `it->ruleInfo == ruleInfo` and avoid a dereference.